// unenv/runtime/node/console/$cloudflare.mjs
export {
  Console,
  _ignoreErrors,
  _stderr,
  _stderrErrorHandler,
  _stdout,
  _stdoutErrorHandler,
  _times
} from "./index.mjs";
import {
  Console as Console2,
  _ignoreErrors as _ignoreErrors2,
  _stderr as _stderr2,
  _stderrErrorHandler as _stderrErrorHandler2,
  _stdout as _stdout2,
  _stdoutErrorHandler as _stdoutErrorHandler2,
  _times as _times2
} from "./index.mjs";
const workerdConsole = globalThis["console"];
export const {
  assert,
  clear,
  // @ts-expect-error undocumented public API
  context,
  count,
  countReset,
  // @ts-expect-error undocumented public API
  createTask,
  debug,
  dir,
  dirxml,
  error,
  group,
  groupCollapsed,
  groupEnd,
  info,
  log,
  profile,
  profileEnd,
  table,
  time,
  timeEnd,
  timeLog,
  timeStamp,
  trace,
  warn
} = workerdConsole;
Object.assign(workerdConsole, {
  Console: Console2,
  _ignoreErrors: _ignoreErrors2,
  _stderr: _stderr2,
  _stderrErrorHandler: _stderrErrorHandler2,
  _stdout: _stdout2,
  _stdoutErrorHandler: _stdoutErrorHandler2,
  _times: _times2
});
export default workerdConsole;
