// unenv/runtime/node/url/internal/constants.cjs
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.CHAR_ZERO_WIDTH_NOBREAK_SPACE = exports.CHAR_VERTICAL_LINE = exports.CHAR_UPPERCASE_Z = exports.CHAR_UPPERCASE_C = exports.CHAR_UPPERCASE_A = exports.CHAR_UNDERSCORE = exports.CHAR_TAB = exports.CHAR_SPACE = exports.CHAR_SINGLE_QUOTE = exports.CHAR_SEMICOLON = exports.CHAR_RIGHT_SQUARE_BRACKET = exports.CHAR_RIGHT_CURLY_BRACKET = exports.CHAR_RIGHT_ANGLE_BRACKET = exports.CHAR_QUESTION_MARK = exports.CHAR_PLUS = exports.CHAR_PERCENT = exports.CHAR_NO_BREAK_SPACE = exports.CHAR_LOWERCASE_Z = exports.CHAR_LOWERCASE_N = exports.CHAR_LOWERCASE_E = exports.CHAR_LOWERCASE_B = exports.CHAR_LOWERCASE_A = exports.CHAR_LINE_FEED = exports.CHAR_LEFT_SQUARE_BRACKET = exports.CHAR_LEFT_CURLY_BRACKET = exports.CHAR_LEFT_ANGLE_BRACKET = exports.CHAR_HYPHEN_MINUS = exports.CHAR_HASH = exports.CHAR_GRAVE_ACCENT = exports.CHAR_FORWARD_SLASH = exports.CHAR_FORM_FEED = exports.CHAR_EXCLAMATION_MARK = exports.CHAR_EQUAL = exports.CHAR_DOUBLE_QUOTE = exports.CHAR_DOT = exports.CHAR_COLON = exports.CHAR_CIRCUMFLEX_ACCENT = exports.CHAR_CARRIAGE_RETURN = exports.CHAR_BACKWARD_SLASH = exports.CHAR_AT = exports.CHAR_AMPERSAND = exports.CHAR_9 = exports.CHAR_0 = void 0;
const CHAR_UPPERCASE_A = exports.CHAR_UPPERCASE_A = 65;
const CHAR_LOWERCASE_A = exports.CHAR_LOWERCASE_A = 97;
const CHAR_UPPERCASE_Z = exports.CHAR_UPPERCASE_Z = 90;
const CHAR_LOWERCASE_Z = exports.CHAR_LOWERCASE_Z = 122;
const CHAR_UPPERCASE_C = exports.CHAR_UPPERCASE_C = 67;
const CHAR_LOWERCASE_B = exports.CHAR_LOWERCASE_B = 98;
const CHAR_LOWERCASE_E = exports.CHAR_LOWERCASE_E = 101;
const CHAR_LOWERCASE_N = exports.CHAR_LOWERCASE_N = 110;
const CHAR_DOT = exports.CHAR_DOT = 46;
const CHAR_FORWARD_SLASH = exports.CHAR_FORWARD_SLASH = 47;
const CHAR_BACKWARD_SLASH = exports.CHAR_BACKWARD_SLASH = 92;
const CHAR_VERTICAL_LINE = exports.CHAR_VERTICAL_LINE = 124;
const CHAR_COLON = exports.CHAR_COLON = 58;
const CHAR_QUESTION_MARK = exports.CHAR_QUESTION_MARK = 63;
const CHAR_UNDERSCORE = exports.CHAR_UNDERSCORE = 95;
const CHAR_LINE_FEED = exports.CHAR_LINE_FEED = 10;
const CHAR_CARRIAGE_RETURN = exports.CHAR_CARRIAGE_RETURN = 13;
const CHAR_TAB = exports.CHAR_TAB = 9;
const CHAR_FORM_FEED = exports.CHAR_FORM_FEED = 12;
const CHAR_EXCLAMATION_MARK = exports.CHAR_EXCLAMATION_MARK = 33;
const CHAR_HASH = exports.CHAR_HASH = 35;
const CHAR_SPACE = exports.CHAR_SPACE = 32;
const CHAR_NO_BREAK_SPACE = exports.CHAR_NO_BREAK_SPACE = 160;
const CHAR_ZERO_WIDTH_NOBREAK_SPACE = exports.CHAR_ZERO_WIDTH_NOBREAK_SPACE = 65279;
const CHAR_LEFT_SQUARE_BRACKET = exports.CHAR_LEFT_SQUARE_BRACKET = 91;
const CHAR_RIGHT_SQUARE_BRACKET = exports.CHAR_RIGHT_SQUARE_BRACKET = 93;
const CHAR_LEFT_ANGLE_BRACKET = exports.CHAR_LEFT_ANGLE_BRACKET = 60;
const CHAR_RIGHT_ANGLE_BRACKET = exports.CHAR_RIGHT_ANGLE_BRACKET = 62;
const CHAR_LEFT_CURLY_BRACKET = exports.CHAR_LEFT_CURLY_BRACKET = 123;
const CHAR_RIGHT_CURLY_BRACKET = exports.CHAR_RIGHT_CURLY_BRACKET = 125;
const CHAR_HYPHEN_MINUS = exports.CHAR_HYPHEN_MINUS = 45;
const CHAR_PLUS = exports.CHAR_PLUS = 43;
const CHAR_DOUBLE_QUOTE = exports.CHAR_DOUBLE_QUOTE = 34;
const CHAR_SINGLE_QUOTE = exports.CHAR_SINGLE_QUOTE = 39;
const CHAR_PERCENT = exports.CHAR_PERCENT = 37;
const CHAR_SEMICOLON = exports.CHAR_SEMICOLON = 59;
const CHAR_CIRCUMFLEX_ACCENT = exports.CHAR_CIRCUMFLEX_ACCENT = 94;
const CHAR_GRAVE_ACCENT = exports.CHAR_GRAVE_ACCENT = 96;
const CHAR_AT = exports.CHAR_AT = 64;
const CHAR_AMPERSAND = exports.CHAR_AMPERSAND = 38;
const CHAR_EQUAL = exports.CHAR_EQUAL = 61;
const CHAR_0 = exports.CHAR_0 = 48;
const CHAR_9 = exports.CHAR_9 = 57;
