// unenv/runtime/node/timers/$cloudflare.cjs
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
Object.defineProperty(exports, "_unrefActive", {
  enumerable: true,
  get: function () {
    return _index._unrefActive;
  }
});
Object.defineProperty(exports, "active", {
  enumerable: true,
  get: function () {
    return _index.active;
  }
});
Object.defineProperty(exports, "clearImmediate", {
  enumerable: true,
  get: function () {
    return _setImmediate.clearImmediateFallback;
  }
});
Object.defineProperty(exports, "clearInterval", {
  enumerable: true,
  get: function () {
    return _index.clearInterval;
  }
});
Object.defineProperty(exports, "clearTimeout", {
  enumerable: true,
  get: function () {
    return _index.clearTimeout;
  }
});

Object.defineProperty(exports, "enroll", {
  enumerable: true,
  get: function () {
    return _index.enroll;
  }
});
Object.defineProperty(exports, "promises", {
  enumerable: true,
  get: function () {
    return _index.promises;
  }
});
Object.defineProperty(exports, "setImmediate", {
  enumerable: true,
  get: function () {
    return _setImmediate.setImmediateFallback;
  }
});
Object.defineProperty(exports, "setInterval", {
  enumerable: true,
  get: function () {
    return _index.setInterval;
  }
});
Object.defineProperty(exports, "setTimeout", {
  enumerable: true,
  get: function () {
    return _index.setTimeout;
  }
});
Object.defineProperty(exports, "unenroll", {
  enumerable: true,
  get: function () {
    return _index.unenroll;
  }
});
var _index = require("./index.cjs");
var _setImmediate = require("./internal/set-immediate.cjs");
module.exports = {
  _unrefActive: _index._unrefActive,
  active: _index.active,
  clearImmediate: _setImmediate.clearImmediateFallback,
  clearInterval: _index.clearInterval,
  clearTimeout: _index.clearTimeout,
  enroll: _index.enroll,
  promises: _index.promises,
  setImmediate: _setImmediate.setImmediateFallback,
  setInterval: _index.setInterval,
  setTimeout: _index.setTimeout,
  unenroll: _index.unenroll
};
