// unenv/runtime/node/assert/strict/index.cjs
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
Object.defineProperty(exports, "AssertionError", {
  enumerable: true,
  get: function () {
    return _index.AssertionError;
  }
});
Object.defineProperty(exports, "CallTracker", {
  enumerable: true,
  get: function () {
    return _index.CallTracker;
  }
});
Object.defineProperty(exports, "deepEqual", {
  enumerable: true,
  get: function () {
    return _index.deepStrictEqual;
  }
});
Object.defineProperty(exports, "deepStrictEqual", {
  enumerable: true,
  get: function () {
    return _index.deepStrictEqual;
  }
});

Object.defineProperty(exports, "doesNotMatch", {
  enumerable: true,
  get: function () {
    return _index.doesNotMatch;
  }
});
Object.defineProperty(exports, "doesNotReject", {
  enumerable: true,
  get: function () {
    return _index.doesNotReject;
  }
});
Object.defineProperty(exports, "doesNotThrow", {
  enumerable: true,
  get: function () {
    return _index.doesNotThrow;
  }
});
Object.defineProperty(exports, "equal", {
  enumerable: true,
  get: function () {
    return _index.strictEqual;
  }
});
Object.defineProperty(exports, "fail", {
  enumerable: true,
  get: function () {
    return _index.fail;
  }
});
Object.defineProperty(exports, "ifError", {
  enumerable: true,
  get: function () {
    return _index.ifError;
  }
});
Object.defineProperty(exports, "match", {
  enumerable: true,
  get: function () {
    return _index.match;
  }
});
Object.defineProperty(exports, "notDeepEqual", {
  enumerable: true,
  get: function () {
    return _index.notDeepStrictEqual;
  }
});
Object.defineProperty(exports, "notDeepStrictEqual", {
  enumerable: true,
  get: function () {
    return _index.notDeepStrictEqual;
  }
});
Object.defineProperty(exports, "notEqual", {
  enumerable: true,
  get: function () {
    return _index.notStrictEqual;
  }
});
Object.defineProperty(exports, "notStrictEqual", {
  enumerable: true,
  get: function () {
    return _index.notStrictEqual;
  }
});
Object.defineProperty(exports, "ok", {
  enumerable: true,
  get: function () {
    return _index.ok;
  }
});
Object.defineProperty(exports, "rejects", {
  enumerable: true,
  get: function () {
    return _index.rejects;
  }
});
Object.defineProperty(exports, "strict", {
  enumerable: true,
  get: function () {
    return _index.strict;
  }
});
Object.defineProperty(exports, "strictEqual", {
  enumerable: true,
  get: function () {
    return _index.strictEqual;
  }
});
Object.defineProperty(exports, "throws", {
  enumerable: true,
  get: function () {
    return _index.throws;
  }
});
var _index = require("../index.cjs");
module.exports = Object.assign(_index.ok, {
  AssertionError: _index.AssertionError,
  CallTracker: _index.CallTracker,
  strict: _index.strict,
  fail: _index.fail,
  ok: _index.ok,
  throws: _index.throws,
  rejects: _index.rejects,
  doesNotThrow: _index.doesNotThrow,
  doesNotReject: _index.doesNotReject,
  ifError: _index.ifError,
  match: _index.match,
  doesNotMatch: _index.doesNotMatch,
  notDeepStrictEqual: _index.notDeepStrictEqual,
  notDeepEqual: _index.notDeepStrictEqual,
  strictEqual: _index.strictEqual,
  equal: _index.strictEqual,
  notStrictEqual: _index.notStrictEqual,
  notEqual: _index.notStrictEqual,
  deepStrictEqual: _index.deepStrictEqual,
  deepEqual: _index.deepStrictEqual
});
