// unenv/runtime/node/events/index.cjs
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
Object.defineProperty(exports, "EventEmitter", {
  enumerable: true,
  get: function () {
    return _events._EventEmitter;
  }
});
Object.defineProperty(exports, "EventEmitterAsyncResource", {
  enumerable: true,
  get: function () {
    return _events.EventEmitterAsyncResource;
  }
});
Object.defineProperty(exports, "addAbortListener", {
  enumerable: true,
  get: function () {
    return _events.addAbortListener;
  }
});
exports.errorMonitor = exports.defaultMaxListeners = exports.default = exports.captureRejections = exports.captureRejectionSymbol = void 0;
Object.defineProperty(exports, "getEventListeners", {
  enumerable: true,
  get: function () {
    return _events.getEventListeners;
  }
});
Object.defineProperty(exports, "getMaxListeners", {
  enumerable: true,
  get: function () {
    return _events.getMaxListeners;
  }
});
exports.listenerCount = exports.init = void 0;
Object.defineProperty(exports, "on", {
  enumerable: true,
  get: function () {
    return _events.on;
  }
});
Object.defineProperty(exports, "once", {
  enumerable: true,
  get: function () {
    return _events.once;
  }
});
exports.usingDomains = exports.setMaxListeners = void 0;
var _events = require("./internal/events.cjs");
const usingDomains = exports.usingDomains = _events._EventEmitter.usingDomains;
const captureRejectionSymbol = exports.captureRejectionSymbol = _events._EventEmitter.captureRejectionSymbol;
const captureRejections = exports.captureRejections = false;
const errorMonitor = exports.errorMonitor = _events._EventEmitter.errorMonitor;
const defaultMaxListeners = exports.defaultMaxListeners = _events._EventEmitter.defaultMaxListeners;
const setMaxListeners = exports.setMaxListeners = _events._EventEmitter.setMaxListeners;
const listenerCount = exports.listenerCount = _events._EventEmitter.listenerCount;
const init = exports.init = _events._EventEmitter.init;
module.exports = _events._EventEmitter;
